components/greaselion/browser/greaselion_download_service.cc

```diff
@@ -0,0 +1,209 @@
+      urls_.ClearPatterns();
+      return;
+    }
+    urls_.AddPattern(pattern);
+  }
+  scoped_refptr<base::SequencedTaskRunner> task_runner =
+      base::CreateSequencedTaskRunnerWithTraits(
+          {base::MayBlock(), base::TaskPriority::USER_VISIBLE,
+           base::TaskShutdownBehavior::SKIP_ON_SHUTDOWN});
+  for (const auto& scripts_it : scripts_value->GetList()) {
+    base::FilePath script_path =
+        root_dir.AppendASCII(kGreaselionConfigFileVersion)
+            .AppendASCII(scripts_it.GetString());
+    if (script_path.ReferencesParent()) {
+      LOG(ERROR) << "Malformed filename in Greaselion configuration";
+    } else {
+      // Read script file on task runner to avoid file I/O on main thread.
+      auto script_contents = std::make_unique<std::string>();
+      std::string* buffer = script_contents.get();
+      base::PostTaskAndReplyWithResult(
+          task_runner.get(), FROM_HERE,
+          base::BindOnce(&base::ReadFileToString, script_path, buffer),
+          base::BindOnce(&GreaselionRule::AddScriptAfterLoad,
+                         weak_factory_.GetWeakPtr(),
+                         std::move(script_contents)));
+    }
+  }
+}
+
+GreaselionRule::~GreaselionRule() = default;
+
+void GreaselionRule::AddScriptAfterLoad(std::unique_ptr<std::string> contents,
+                                        bool did_load) {
+  if (!did_load || !contents) {
+    LOG(ERROR) << "Could not load Greaselion script";
+    return;
+  }
+  scripts_.push_back(*contents);
+}
+
+bool GreaselionRule::PreconditionFulfilled(
+    GreaselionPreconditionValue precondition,
+    bool value) const {
+  switch (precondition) {
+    case kMustBeTrue:
+      return value;
+    case kMustBeFalse:
+      return !value;
+    default:
+      return true;
+  }
+}
+
+bool GreaselionRule::Matches(const GURL& url, GreaselionFeatures state) const {
+  if (!PreconditionFulfilled(preconditions_.rewards_enabled,
+                             state[greaselion::REWARDS]))
+    return false;
+  if (!PreconditionFulfilled(preconditions_.twitter_tips_enabled,
+                             state[greaselion::TWITTER_TIPS]))
+    return false;
+  return urls_.MatchesURL(url);
+}
+
+void GreaselionRule::Populate(std::vector<std::string>* scripts) const {
+  scripts->insert(scripts->end(), scripts_.begin(), scripts_.end());
+}
+
+GreaselionDownloadService::GreaselionDownloadService(
+    LocalDataFilesService* local_data_files_service)
+    : LocalDataFilesObserver(local_data_files_service),
+      task_runner_(base::CreateSequencedTaskRunnerWithTraits(
+          {base::MayBlock(), base::TaskPriority::USER_VISIBLE,
+           base::TaskShutdownBehavior::SKIP_ON_SHUTDOWN})),
+      weak_factory_(this) {
+  DETACH_FROM_SEQUENCE(sequence_checker_);
+}
+
+GreaselionDownloadService::~GreaselionDownloadService() {}
+
+void GreaselionDownloadService::OnDATFileDataReady(std::string contents) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
```

Comment:
this is called on the reply so should be UI thread, right? We should make sure that every method that references `rules_` has a `DCHECK` to make sure that it's called on the same thread

Reply:
Added `DCHECK_CALLED_ON_VALID_SEQUENCE` to `rules()`